src/components/shared/MarkerTooltipContents.js

```diff
@@ -96,18 +96,90 @@ function _makePhaseTimesArray(
  * For example, it isn't useful to say "sweeping took 200ms" but it is
  * useful to say "sweeping.compacting took 20ms" (if compacting has no
  * sub-phases.
+ *
+ * We find leaf phases by constructing a tree of phase times and then
+ * reading its leaves.
  */
-function _isLeafPhase(
-  phaseName: string,
-  phases: Array<PhaseTimeTuple>
-): boolean {
-  return phases.every(
-    pt =>
-      // This phase is not the start of some other phase
-      !pt.name.startsWith(phaseName) ||
-      // Except itself.
-      pt.name === phaseName
-  );
+
+type PhaseTreeNode = PhaseTreeLeafNode | PhaseTreeBranchNode;
+
+type PhaseTreeLeafNode = {
+  node_type: 'leaf',
+  value: PhaseTimeTuple,
+};
+
+type PhaseTreeBranchNode = {
+  node_type: 'branch',
+  branches: Map<string, PhaseTreeNode>,
+};
+
+function _treeInsert(
+  tree: Map<string, PhaseTreeNode>,
+  path: Array<string>,
+  phase: PhaseTimeTuple
+) {
+  const component = path.shift();
+  if (component === undefined) {
+    // This path is not a leaf, it can be ignored.
+  } else {
+    const node = tree.get(component);
+    if (node) {
```

Comment:
I think the branches diverge a little too much to do that, at least with this update-in-place style I've picked.

Reply:
I think that what I'm confused with is that I don't get what is different in the code I've put above (NOTE: I slightly updated it by nullifying node.value in the "branch" branch as I missed it initially).

I think it's nearly the same behavior but is easier to follow. A difference is the "duplicate phase" case that's silently ignored, but we shouldn't throw an error anyway because we're actually at the `render` phase, and throwing is really a bad idea here.

If I miss something and it isn't the same behavior I'm fine keeping your code, with the "throw" replaced by a "console.error" or just silently ignoring (sorry for missing it earlier...), and an early return for the "undefined" case.

Reply:
Hey @PaulBone, the last point that I wanted to check with you is what's in this comment here ^.

Just to reinstate what I said earlier, I'm trying to make this piece of code easier to read and maintain. I try to understand if my piece of code above has the same behavior as yours. I think it does.

Maybe we'd need some tests for this to check it :) I can invest some time in it once I'm done with my current work with the sidebar if necessary.

Reply:
Hi @julienw you're right, it can be written like this.  I persionally think it's less readable. but years of functional programming has done something to my brain and I natrually follow structual induction or other types of induction when writing these things, and avoid early returns and other things.